File: 54. Spiral Matrix/54. Spiral Matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
  const ans = [];
  if (matrix.length == 0) return ans;

  let r1 = 0,
    r2 = matrix.length - 1;
  let c1 = 0,
    c2 = matrix[0].length - 1;
  while (r1 <= r2 && c1 <= c2) {
    for (let j = c1; j <= c2; j++) ans.push(matrix[r1][j]);
    for (let i = r1 + 1; i <= r2; i++) ans.push(matrix[i][c2]);
    if (r1 == r2 || c1 == c2) break;
    for (j = c2 - 1; j > c1; j--) ans.push(matrix[r2][j]);
    for (i = r2; i > r1; i--) ans.push(matrix[i][c1]);
    r1++;
    r2--;
    c1++;
    c2--;
  }
  return ans;
};

File: 54. Spiral Matrix/54. Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        if(matrix.length == 0) return ans;
        
        int R = matrix.length;
        int C = matrix[0].length;
        boolean[][]seen = new boolean[R][C];
        int[]dr = {0, 1, 0, -1};
        int[]dc = {1, 0, -1, 0};
        
        int c = 0, r = 0, di = 0;
        for(int i = 0; i < R * C; i++){
            ans.add(matrix[r][c]);
            seen[r][c] = true;
            int nextr = r + dr[di];
            int nextc = c + dc[di];
            
            if(nextr >= 0 && nextr < R && nextc >= 0 && nextc < C && !seen[nextr][nextc]){
                r = nextr;
                c = nextc;
            }else{
                di = (di + 1) % 4;
                r += dr[di];
                c += dc[di];
            }
           
        }
         return ans;
    }
}